src/fsharp/FSharp.Core/local.fs

```diff
@@ -85,6 +85,42 @@ module internal List =
                 cons <- cons2
             setFreshConsTail cons []
             res
+      
+    let inline groupBy (comparer:IEqualityComparer<'SafeKey>) (keyf:'T->'SafeKey) (getKey:'SafeKey->'Key) (list: 'T list) =
+        let dict = Dictionary<_, _ list []> comparer
+
+        // Build the groupings
+        let rec loop list =
+            match list with
+            | v :: t -> 
+                let safeKey = keyf v
+                match dict.TryGetValue(safeKey) with
+                | true, prev -> 
+                       let cons2 = freshConsNoTail v
+                       setFreshConsTail prev.[1] cons2
+                       prev.[1] <- cons2
+                | _ -> let res = freshConsNoTail v
+                       dict.[safeKey] <- [|res; res |] // First index stores the result list; second index is the most recent cons.
+                    
+                loop t
+            | _ -> ()
+        loop list  
+
+        let mutable ie = dict.GetEnumerator()
+        if not (ie.MoveNext()) then []
+        else
+            let finalCons = ie.Current.Value.[1]
+            setFreshConsTail finalCons []
+            let res = freshConsNoTail (getKey ie.Current.Key, ie.Current.Value.[0])
+            let mutable cons = res
+            while ie.MoveNext() do
+                let finalCons = ie.Current.Value.[1]
```

Comment:
You evaluate ``ie.Current`` three times on each iteration - please evaluate it only once.